### PR TITLE
Add parallel matching for `tf.gfile.Glob`

### PR DESCRIPTION
**Pull request by firejq**
This is a PR from JIZHI Team & TaiJi AI platform in Tencent.

Add a argument `max_parallel_num` for  `tf.gfile.Glob`, which will setup a threadpool to parallelize the matching for mulitple pattern strings. It can bring nice performance improvement if `max_parallel_num` is set appropriately.

Thanks for your review.

### CODE REVIEW

**Comment by mihaimaruseac** [comment]
Since the matching happensinside C++ code, I'm afraid this PR does not do what it says it does.

**Comment by firejq** [comment]
@mihaimaruseac Thanks for your reply! 

I also considered the issue you mentioned in fact. In the C++ implement of different `FileSystem`, there may be some mutexes which cause the multi-threaded parallel from python client does not work. However, `glob` is a I/O-bound operation as we known, which always consumes a lot of time in blocking wait and multi-threads can ease this overhead a lot.

For reference, I have obtained nice effect in my case that using `tf.gfile.Glob` for hdfs:
```python
pattern = ['hdfs://xxxx_1/*', 'hdfs://xxxx_2/', 'hdfs://xxxx_3/, 'hdfs://xxxx_4/', 'hdfs://xxxx_5/']
begin_time = int(time.time())
files = tf.gfile.Glob(pattern, max_parallel_num=5)
print('size: {}, used time: {} s'.format(len(files), int(time.time()) - begin_time))
```
As the final size of `files` is 20000, the above code costed 146 seconds as it printed, and 809 seconds while `max_parallel_num` is set to `1` (It is too slow for me so I have to make some optimization...).

The above is my reason for proposing this pr and there may be some incompleteness. I am looking forward to your valuable suggestions for improvement.

**Comment by mihaimaruseac** [comment]
This looks good to me now. Since it changes API, let's wait for API owners approval too

File: tensorflow/python/lib/io/file_io.py
Comment (fchollet) [inline tensorflow/python/lib/io/file_io.py:343]: This should be `num_threads` (the max mention is self-evident and thus not necessary)

Note: `num_threads` / `num_workers` is already an argument in several places.

File: tensorflow/python/lib/io/file_io.py
Comment (fchollet) [inline tensorflow/python/lib/io/file_io.py:444]: Some concern about the performance of this implementation. This should be taking a list of patterns

Also: please provide a benchmark for this implementation

**Review by fchollet — COMMENTED** [review]
Maybe the choice of the number of threads to use should be automated, rather than specified by the user. Most users will not be aware of this argument. What do you think?

**Comment by firejq** [comment]
@fchollet Thanks for your review!

> Maybe the choice of the number of threads to use should be automated, rather than specified by the user. Most users will not be aware of this argument. What do you think?

In fact I have implemented automatically using `len(pattern)` as the number of parallel processes inside the api, however, I overturned my own ideas in my practice. As for the cases which run multiple tf instance processes on the same machine (e.g. tf with horovod), users tend to control the parallel number of different processes based on the CPU resources on their machine. If the api automatically tries the maximum parallelism, it may cause unexpected resource preemption. 

This is my reason for setting it as a configurable parameter and I would like to discuss further with you.

File: tensorflow/python/lib/io/file_io.py
Comment (firejq) [reply tensorflow/python/lib/io/file_io.py]: All right! Done.

File: tensorflow/python/lib/io/file_io.py
Comment (firejq) [reply tensorflow/python/lib/io/file_io.py:444]: Each thread does return a list in this implementation and so I used `functools.reduce` to flatten these lists before returning the result. I think the answer [here](https://stackoverflow.com/a/45323085/6072681) is a valuable reference as for benchmark, which is why I chose it.

**Comment by firejq** [comment]
@fchollet Could you please check this again?

File: tensorflow/python/lib/io/file_io.py
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.py:440]: I would move this to the beginning of the function and use it in the `if isinstance(pattern, six.string_types):` case too.

File: tensorflow/python/lib/io/file_io.py
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.py:431]: Inside this if, we should just set `num_threads` to 1 / raise error if it is strictly below 1.

Then we can reuse the pool implementation, it should work with a single thread too, right?

File: tensorflow/python/lib/io/file_io.py
Comment (mihaimaruseac) [inline tensorflow/python/lib/io/file_io.py:364]: I thought we were going to automatically determine the best number of threads if user didn't supply any.

**Review by mihaimaruseac — COMMENTED** [review]
Please don't forget about posting some benchmark results

File: tensorflow/python/lib/io/file_io.py
Comment (firejq) [reply tensorflow/python/lib/io/file_io.py]: Have done.

File: tensorflow/python/lib/io/file_io.py
Comment (firejq) [reply tensorflow/python/lib/io/file_io.py]: All right! I have modified as you said. 

The reason for handling the case of one separately is due to the overhead of the thread pool created, but this overhead indeed accounts for a small proportion in the current situation.

File: tensorflow/python/lib/io/file_io.py
Comment (firejq) [reply tensorflow/python/lib/io/file_io.py]: Have done.

**Comment by firejq** [comment]
> Please don't forget about posting some benchmark results

@mihaimaruseac Sorry for my delayed benchmark..

System environment: Intel(R) Xeon(R) CPU E5-2699 v4 @ 2.20GHz 88 threads

I used the following 5 different directories of posix file sytem for benchmarking, the first of which is an empty directory to observe the overhead of the operation itself:

- d1: a empty directory
- d2: 4912 directories, 44711 files
- d3: 7601 directories, 76354 files
- d4: 64520 directories, 905496 files
- d5: 62456 directories, 881247 files

Each test is repeated 10 times and the average is taken as the result. The all benchmark results are as follows:

| num_threads   | patterns    | Duration (ms) |
| ------------- | ----------- | ------------- |
| 1             | [ ]         | 130           |
| 1             | d1          | 372           |
| 1             | d1+d2       | 6635          |
| 1             | d1+d2+d3    | 10355         |
| 1             | d1+d2+d3+d4 | 50824         |
| 1             | d1+d2+d3+d4+d5 | 88656         |
| len(patterns) | [ ]         | 142           |
| len(patterns) | d1          | 395           |
| len(patterns) | d1+d2       | 6719          |
| len(patterns) | d1+d2+d3    | 7883          |
| len(patterns) | d1+d2+d3+d4 | 38811         |
| len(patterns) | d1+d2+d3+d4+d5 | 48539         |

**Comment by mihaimaruseac** [comment]
So this does not seem to bring a huge speed gain and is a slowdown in the first 3 scenarios. Can you run the benchmark again after #44310 landed?

**Comment by firejq** [comment]
@mihaimaruseac 

Since the parallelism of this pr is for multiple different patterns in Python, only when multiple patterns are input can the corresponding optimization effect be reflected. Therefore, the benchmarks here mainly use the number of input patterns as experimental variables.

> So this does not seem to bring a huge speed gain and is a slowdown in the first 3 scenarios.

For the test results above,
- The increased duration of the first scenario should be caused by other random factors, because when the input is `[ ]`, it returns directly [here](https://github.com/tensorflow/tensorflow/blob/a8a7a1cd5d5131fb0e409f4c919530d61ec3646e/tensorflow/python/lib/io/file_io.py#L432) without subsequent performing. In fact, the reason I set this is to take into account some random overhead.
- For the second and third results, due to the small number of patterns, I think the [overhead of pool](https://github.com/tensorflow/tensorflow/pull/44269#discussion_r516407046) is the main reason for the increased duration. 

> Can you run the benchmark again after #44310 landed?

I have made benchmarks again with #44310 landed in the same environment, and the results are as follows:

| num_threads   | patterns    | Duration (ms) |
| ------------- | ----------- | ------------- |
| 1             | [ ]         | 428           |
| 1             | [d1]          | 489           |
| 1             | [d1+d2]       | 1,796          |
| 1             | [d1+d2+d3]    | 4,194         |
| 1             | [d1+d2+d3+d4] | 16,353         |
| 1             | [d1+d2+d3+d4+d5] | 29,256         |
| len(patterns) | [ ]         | 425           |
| len(patterns) | [d1]          |  496          |
| len(patterns) | [d1+d2]       | 1,654          |
| len(patterns) | [d1+d2+d3]    | 3,216          |
| len(patterns) | [d1+d2+d3+d4] | 11,622         |
| len(patterns) | [d1+d2+d3+d4+d5] | 19,753         |

Considering that this increase is actually acceptable for this `Glob` api (right?), and when the number of patterns increases, the effect of parallel optimization can be reflected, we have reason to think this parallel in python is still quite valuable.

File: tensorflow/python/lib/io/file_io.py
Comment (fchollet) [inline tensorflow/python/lib/io/file_io.py:347]: What happens with the default value of None?

**Comment by mihaimaruseac** [comment]
This is awesome. Let's add a comment saying what the default value of the `num_threads` is (when user passes `None`/doesn't pass anything). And then we can merge.

**Comment by firejq** [comment]
@mihaimaruseac @fchollet  Thanks for your review! I have updated the comment about the default value of `num_threads`.

File: tensorflow/python/lib/io/file_io.py
Comment (firejq) [reply tensorflow/python/lib/io/file_io.py]: Sorry that I reopened this conversation. As for the default values of `num_threads`, I think we may need to discuss it again.

I have made [benchmarks](https://github.com/tensorflow/tensorflow/pull/44269#issuecomment-721501440) on a multi-core (80 threads) server and the default value of `num_threads` (`len(patterns)`) can produce a quite positive optimization effect. However, when I apply this to another machine with poorer performance, it may produces negative effect and even sometimes makes the process hang. I think the cause is excessive CPU resource preemption with the parallelism enhanced at the C++ level (#44310).

**Therefore, considering of better versatility, I prefer to set the default value of `num_threads` to `1`**, i.e. turn [here](https://github.com/tensorflow/tensorflow/blob/d8192d06e91ce3be15e5ff8b95641bfe76226803/tensorflow/python/lib/io/file_io.py#L435) to `num_threads = num_threads or 1`, which can keep consistent with the previous codes by default, and work positively if user set it manually. In addition, we can explain the risk and recommended value of this parameter in the comments.

Look forward to your suggestions.

**Comment by mihaimaruseac** [comment]
I would also go with 1 thread by default to maintain compatibility. Alternatively, minimum between number of patterns and number of cores.

**Comment by firejq** [comment]
@mihaimaruseac 
Thanks! I have finished the update of default value setting and I think it has reached a more robust state now.  Could you please check anything need enhancement more again? 😄

**Review by mihaimaruseac — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by firejq** [comment]
Hey @gbaned  There seems to be a internal failed check in the ci process. Could you please have a look?

**Comment by firejq** [comment]
Hi, @mihaimaruseac @gbaned 

There seems to be a internal failed check in the ci process. Could you please help to see if something went wrong?

**Comment by mihaimaruseac** [comment]
It is imported internally, pending some additional review. I'll come back here if there's more to do, but for now this is all good.

**Comment by firejq** [comment]
> It is imported internally, pending some additional review. I'll come back here if there's more to do, but for now this is all good.

Okay, thanks for pushing forward.

**Comment by mihaimaruseac** [comment]
An API owners comment:

> Can we push this into C++ too? Multiprocessing in Python has been the source of several bugs in the past

I think this should possible. What do you think?

**Comment by firejq** [comment]
> An API owners comment:
> 
> > Can we push this into C++ too? Multiprocessing in Python has been the source of several bugs in the past
> 
> I think this should possible. What do you think?

@mihaimaruseac 
Does he mean that C++ OP `_pywrap_file_io.GetMatchingFiles` directly receives a list of patterns, and then processes the multiple patterns in parallel? 

If so, first of all, I think `push this into C++ too` is worth considering. As we discussed above, [changing the default value to `1`](https://github.com/tensorflow/tensorflow/pull/44269#discussion_r517997186) is to avoid introducing uncertainty from Python multiprocesses in fact. In this pr, python's parallelism is actually chosen by the user according to the specific running environment (just like other `dataset`'s parallelism parameters). The default parameters are always the same value as before and do not pose any uncertainty to the existing user code.

On the other hand, this will also result in more changes, including input formats for C++ Op. So I prefer to preserve python's multiprocesses for now. And of course, if the owner sticks to this point, I can also make the changes, but it may take more time and more disscussion. Thanks anyway!

**Comment by mihaimaruseac** [comment]
We can change C++ API (and update the goldens files), it is not an issue, as long as Python API maintains compatibility.

Seems python multiprocessing has issues so it's preferable to push this code to C++.

**Comment by firejq** [comment]
> We can change C++ API (and update the goldens files), it is not an issue, as long as Python API maintains compatibility.
> 
> Seems python multiprocessing has issues so it's preferable to push this code to C++.

Sorry for the late reply...
Okay,  I will migrate the relevant logic to C++ later. Thanks!

**Comment by gbaned** [comment]
@firejq Can you please resolve conflicts? Thanks!

**Comment by firejq** [comment]
Due to other tighter issues in the near future, I will temporarily close this pr, and reopen it if there are new developments in the future. Thanks to all reviewers!

**Pull request closed without merging**
